src/shogun/mathematics/linalg/LinalgBackendEigen.h

```diff
@@ -28,47 +27,57 @@
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  * POSSIBILITY OF SUCH DAMAGE.
  *
- * Authors: 2016 Pan Deng, Soumyajit De, Viktor Gal
+ * Authors: 2016 Pan Deng, Soumyajit De, Heiko Strathmann, Viktor Gal
  */
 
-#include <shogun/lib/config.h>
 #include <shogun/mathematics/eigen3.h>
-#include <shogun/mathematics/linalg/CPUVector.h>
-
-#ifndef CPUBACKEND_H__
-#define CPUBACKEND_H__
+#include <shogun/mathematics/linalg/LinalgBackendBase.h>
 
-#ifdef HAVE_CXX11
+#ifndef Linalg_Backend_Eigen_H__
+#define Linalg_Backend_Eigen_H__
 
 namespace shogun
 {
 
-/** CPU Backend Class */
-class CPUBackend
+/** @brief linalg methods with EIGEN3 backend */
+class LinalgBackendEigen : public LinalgBackendBase
 {
 public:
+	/** vector dot-product macro */
+	#define BACKEND_GENERIC_DOT(Type) \
+	virtual Type dot(const SGVector<Type>& a, const SGVector<Type>& b) const \
+	{  \
+		return dot_impl(a, b);  \
+	}
 
-	/** Default Constructor */
-	CPUBackend();
+	BACKEND_GENERIC_DOT(float32_t);
+	BACKEND_GENERIC_DOT(float64_t);
+	BACKEND_GENERIC_DOT(int32_t);
 
-	/** Copy constructor */
-	CPUBackend(const CPUBackend& cpubackend);
+	/** @return object name */
+	virtual const char* get_name() const { return "Eigen3"; }
 
+private:
 	/**
-	 * Implementation of Eigen3 vector dot-product that works
-	 * with CPUVectors (SGVectors)
+	 * Method for internal implementation of vector dot-product that works
+	 * with CPU vectors.
 	 *
 	 * @param a first vector
 	 * @param b second vector
 	 * @return the dot product of \f$\mathbf{a}\f$ and \f$\mathbf{b}\f$, represented
 	 * as \f$\sum_i a_i b_i\f$
 	 */
 	template <typename T>
-	T dot(const CPUVector<T> &a, const CPUVector<T> &b) const;
+	T dot_impl(const SGVector<T>& a, const SGVector<T>& b) const
+	{
+		typedef Eigen::Matrix<T, Eigen::Dynamic, 1> VectorXt;
```

Comment:
pls put that somewhere central.
Or even add a method to SGVector that turns it into an eigen3 map (if not yet there)